plugins/inputs/t128_tank/reader.go

```diff
@@ -0,0 +1,354 @@
+	if err != nil {
+		return StartIndex, fmt.Errorf("unable to parse %s as integer: %s", data, err)
+	}
+	return index{value: indexVal}, nil
+}
+
+func (i index) string() string {
+	if i.value == endIndexUint {
+		return endIndexStr
+	}
+	return fmt.Sprintf("%d", i.value)
+}
+
+func (i index) next() index {
+	if i.value == startIndexUint || i.value == endIndexUint {
+		return i
+	}
+	return index{value: i.value + 1}
+}
+
+type CommandContext = func(ctx context.Context, name string, arg ...string) *exec.Cmd
+
+type Reader struct {
+	topic          string
+	lastSavedIndex uint64
+	// used to send events from the read routine to the send routine
+	sendChan chan IndexedMessage
+	// the target address for the TANK instance
+	tankAddress string
+	// the target port for the TANK instance
+	tankPort int
+	// time to wait before restarting the tank read command
+	restartDelay time.Duration
+	// used to notify the main routine that the read routine finished with an error
+	readDone chan error
+	// used to mock tank read subprocess in testing
+	tankReadCmdCtx CommandContext
+	// path to the index file
+	indexPath string
+	// default index in case reading to/from the index file fails
+	defaultIndex index
+}
+
+func NewReader(tankAddress string, tankPort int, topic string, indexPath string, defaultIndex index) *Reader {
+	return &Reader{
+		topic:          topic,
+		tankReadCmdCtx: exec.CommandContext,
+		sendChan:       make(chan IndexedMessage),
+		readDone:       make(chan error),
+		restartDelay:   5 * time.Second,
+		tankAddress:    tankAddress,
+		tankPort:       tankPort,
+		indexPath:      indexPath,
+		defaultIndex:   defaultIndex,
+	}
+}
+
+func (r *Reader) withTankReadCommandContext(tankReadCmdCtx CommandContext) *Reader {
+	r.tankReadCmdCtx = tankReadCmdCtx
+	return r
+}
+
+func (r *Reader) Run(mainCtx context.Context, stopchan chan struct{}) {
+	readCtx, readCtxCancel := context.WithCancel(mainCtx)
+	paused := false
+	lastIndex, err := getIndex(r.indexPath, r.defaultIndex)
+	r.lastSavedIndex = lastIndex.value
+	if err != nil {
+		log.Printf("Error in get index")
+		return
+	}
+
+	nextSaveCheck := time.After(2 * time.Second)
+	defer func() {
+		readCtxCancel()
+		setIndex(r.indexPath, index{value: r.lastSavedIndex})
+	}()
+
+	go r.read(mainCtx, readCtx, lastIndex.next())
+
+	for {
+		select {
+		case <-mainCtx.Done():
+			fmt.Printf("%s reader done", r.topic)
+			readCtxCancel()
+			return
+		case <-nextSaveCheck:
+			if lastIndex.value > r.lastSavedIndex {
+				setIndex(r.indexPath, lastIndex)
+				r.lastSavedIndex = lastIndex.value
+			}
+			nextSaveCheck = time.After(2 * time.Second)
+		case err := <-r.readDone:
+			if !paused {
+				readCtx, readCtxCancel = context.WithCancel(mainCtx)
+				var boundaryFault *boundaryFault
+				if err != nil && errors.As(err, &boundaryFault) {
+					fmt.Printf("detected boundary fault, restarting %s tank read from index %d", r.topic, boundaryFault.nextAvailableIndex.value)
```

Comment:
telegraf has a logging infrastructure that should be used instead of direct fmt printing

Reply:
yep! All fmt were for my sanity checks.